### PR TITLE
added test artifacts

### LINKED ISSUE #2126: Artifacts property in a test object
Public api improvement. 
This should improve further integration with reporter tools and codecept ui. 

Test object should contain object with artifacts:

```js
test.artifacts = {
  video: pathtovideo, 
  screenshot: pathtoscreenshot, 
  slideshow, source 
} 
```

This should be imemented in screenshot On Fail plugin, and used by Allure plugin.

### PR DESCRIPTION
**Pull request by DavertMik**
## Motivation/Description of the PR

Plugins may attach to a test instance artifacts: screenshots, videos, etc
Reporters like allure, reportportal, others, should be able to collect that artifacts from a test instance.
This PR introduces API for sharing this data.

- Resolves #2126 

To keep backwards compatibility this PR doesn't change behavior of allure plugin, but when tested a code can be simplified by in plugins that manually send attachments to allure.

Applicable helpers:

- [ ] WebDriver
- [ ] Puppeteer
- [ ] Nightmare
- [ ] REST
- [ ] FileHelper
- [ ] Appium
- [ ] Protractor
- [ ] TestCafe
- [ ] Playwright

Applicable plugins:

- [ ] allure
- [ ] autoDelay
- [ ] autoLogin
- [ ] customLocator
- [ ] pauseOnFail
- [ ] puppeteerCoverage
- [ ] retryFailedStep
- [x] screenshotOnFail
- [x] selenoid
- [x] stepByStepReport
- [ ] wdio

## Type of change

- [ ] :fire: Breaking changes
- [ ] :rocket: New functionality
- [ ] :bug: Bug fix
- [ ] :clipboard: Documentation changes/updates
- [ ] :hotsprings: Hot fix
- [ ] :hammer: Markdown files fix - not related to source code
- [ ] :nail_care: Polish code

## Checklist:

- [ ] Tests have been added
- [ ] Documentation has been added (Run `npm run docs`)
- [ ] Lint checking (Run `npm run lint`)
- [ ] Local tests are passed (Run `npm test`)

### CODE REVIEW

**Comment by PeterNgTr** [comment]
Can we have tests for this?

File: lib/listener/artifacts.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+const event = require('../event');
+
+module.exports = function () {
+  event.dispatcher.on(event.all.before, (test) => {
```
Comment (DavertMik) [inline lib/listener/artifacts.js:4]:
```suggestion
  event.dispatcher.on(event.test.before, (test) => {
```

**Comment by DavertMik** [comment]
> Can we have tests for this?

@PeterNgTr Good point but I didn't figure out how to test that properly.

**Comment by DavertMik** [comment]
Selection_841

**Pull request merged**